### ISSUE TITLE
RELEASE: Ypsiliform

### ISSUE THREAD

**Issue by ohrrpgce-bugbot** [ohrrpgce/ohrrpgce#683]
[bz#683]

To make a bug a blocker for the Ypsiliform release, just nominate it here.
To make a a feature request a goal for Ypsiliform, just nominate it here.

From: @bob-the-hamster
Reported version: 20100108 Ypsiliform
Depends on:   bz⁠#738, bz⁠#348, bz⁠#477, bz⁠#646, bz⁠#672, bz⁠#674, bz⁠#684, bz⁠#696, bz⁠#700, bz⁠#704, bz⁠#715, bz⁠#717, bz⁠#731, bz⁠#754, bz⁠#755, bz⁠#762, bz⁠#766, bz⁠#772, bz⁠#774, bz⁠#775, bz⁠#791, bz⁠#793, bz⁠#798, bz⁠#799, bz⁠#801, bz⁠#813, bz⁠#829

**Comment by ohrrpgce-bugbot** [comment]
Comment author: @bob-the-hamster

Does anybody have any more bugs (or feature requests) that they would like to nominate for inclusion in Ypsiliform before I move to start stabilizing for the release?

I would especially like to hear from anybody who has been testing out the plot slice code.

**Comment by ohrrpgce-bugbot** [comment]
Comment author: **Kizul Emeraldfire &lt;KizulEmeraldfire&gt;**

I'd like to nominate bug bz⁠#697:
http://rpg.hamsterrepublic.com/bugzilla/show_bug.cgi?id=697

Seems like it'd be an easy enough thing to fix. :)

**Comment by ohrrpgce-bugbot** [comment]
Comment author: @bob-the-hamster


> I'd like to nominate bug bz⁠#697:
> 
> Seems like it'd be an easy enough thing to fix. :)

To me that one sounds pretty huge and complicated-- but I am willing to at least try.

**Comment by ohrrpgce-bugbot** [comment]
Comment author: @rversteegen

I've added some bugs I think look particularly bad to leave unfixed :)

Plotslices are still pretty unfinished. What about adding a bit more of the plan, such as bug bz⁠#685? Overlays would be so easy, wouldn't that be interesting to see used!

How about adding a 4th map layer: a second overhead (above the current one)? It wouldn't require much, and has been requested several times with good reason.

(Bait and switch)

While we're at that, it only makes sense to implement the related piece of the slices plan and draw map layers as slices and allow rearranging them and other layers. Two overhead layers probably means one used for things like trees and lampposts, and the other for things like weather effects. It's highly likely that you'll want to place sprites beneath the top map layer or two.

**Comment by ohrrpgce-bugbot** [comment]
Comment author: @bob-the-hamster


> I've added some bugs I think look particularly bad to leave unfixed :)
> 
> Plotslices are still pretty unfinished. What about adding a bit more of the
> plan, such as bug bz⁠#685? Overlays would be so easy, wouldn't that be interesting
> to see used!

There is a ton of stuff still to be done with slices, but I am quite happy with the current stuff implemented so far.

> How about adding a 4th map layer: a second overhead (above the current one)? It
> wouldn't require much, and has been requested several times with good reason.
> 
> (Bait and switch)
> 
> While we're at that, it only makes sense to implement the related piece of the
> slices plan and draw map layers as slices and allow rearranging them and other
> layers. Two overhead layers probably means one used for things like trees and
> lampposts, and the other for things like weather effects. It's highly likely
> that you'll want to place sprites beneath the top map layer or two.

Map layers as slices was actually the very next thing on my to-do list :)

**Comment by ohrrpgce-bugbot** [comment]
Comment author: @bob-the-hamster

These crashers are making mes sad :(

**Comment by ohrrpgce-bugbot** [comment]
Comment author: @bob-the-hamster

Because of the script interpreter crash fix in bug bz⁠#717 and because of the large number of new features, I really want to make a new stable release very soon.

It seems to me that all the remaining blockers for this release are just "nice to have" things, and that none of them would represent a step backwards if Ypsiliform released without them.

Thoughts?

**Comment by ohrrpgce-bugbot** [comment]
Comment author: @arperry

A fix for the Vista music crash would be very nice to have :(

FWIW, it's happening with any midi I try (I haven't tried OGG, since it crashed before I got that far) and I ran Baconthulhu (and others, maybe?) on that same computer with no problems.

**Comment by ohrrpgce-bugbot** [comment]
Comment author: @rversteegen

I agree, a release feels overdue when so many people are using and depending on WIP features - though they are still unfinished, which is a pity. At least we have text slices now!

Damn, I promised so many people at least one more map layer for Ypsiliform! I'm going to work on that immediately - tomorrow; there just wouldn't be a slice interface.

**Comment by ohrrpgce-bugbot** [comment]
Comment author: @bob-the-hamster

I am really very much wanting to release Ypsiliform soon. I know we have some bugs remaining, but I don't think any of the open bugs blocking Ypsiliform right now represents an actual step backwards from the Xocolatl release (except maybe bug bz⁠#731 ?)

I have been noticing cases recently where people try to play games written with WIP versions in the Stable release and get confused when they don't work.

**Comment by ohrrpgce-bugbot** [comment]
Comment author: @bob-the-hamster

The mad urge to abruptly and unilaterally declare a stable release is almost overwhelming! My finger is hovering over the big read release-now button! TMC, please talk me down!

**Comment by ohrrpgce-bugbot** [comment]
Comment author: @rversteegen

I can't let you do that, James -- I promised more map layers over and over, and I have to deliver!

Luckily I have a secret weapon - a flurry of release blocking bug reports! Oh, the horror!

**Comment by ohrrpgce-bugbot** [comment]
Comment author: @bob-the-hamster

We need a stable release very very soon.

I am seeing more and more stuff like this: http://castleparadox.com/ohr/viewtopic.php?t=6352

I know the remaining ypsiliform release blockers are important, but they are becoming less important than the data-loss risk from people moving files from Xocolatl to WIP and back to Xocolatl again. I know that is my own fault for leaving the RPG version warning missing/broken in custom for like a million years :(

I don't want to wait much longer than about one more month to release Ypsiliform.

**Comment by ohrrpgce-bugbot** [comment]
Comment author: @rversteegen

This has finally spurred me to add IMPORTANT-nightly.txt.

I don't really consider one month as "very very soon". I'm sure we can get to release candidates quite a bit quicker than that - but they should definitely not be rushed :)

**Comment by ohrrpgce-bugbot** [comment]
Comment author: @raekuul

Pokota is requesting that bug bz⁠#252 (which is actually a feature request) be looked at for Ypsiliform, mainly because it's been three years since filed, and nothing has been seen in Custom (nightlies or otherwise) that indicate progress on it.

**Comment by ohrrpgce-bugbot** [comment]
Comment author: @bob-the-hamster

Sorry for the flood of bug-spam.

All the bugs that I have removed from blocking the release are things I would love to have fixed before the release, but they are not important enough to justify delaying the release further.

Also, in the remaining blockers: bug bz⁠#704, bug bz⁠#738, and bug bz⁠#766 I am doubtful as to whether or not they are worth delaying the release for. They are all very nasty, and I do want them fixed, but when weighed against all the other improvements that are already done, it is hard to justify waiting on them.

And I feel the same way about map layer 4, TMC, hint hint ;)

**Comment by ohrrpgce-bugbot** [comment]
Comment author: @bob-the-hamster

Any objections to a Friday release?

**Comment by ohrrpgce-bugbot** [comment]
Comment author: @rversteegen


> Any objections to a Friday release?

Excuuuse me, I think you mean "Any objections to a Friday release candidate?"

Anyway, yes I do; I've been meaning to post the following, Things To Be Done Before Release:

\-change resetgame
\-change get/setboxborder
 (since these are breaking changes, they really should be done before the next stable release)
\-determine whether the default graphics backend should be changed to gfx_sdl. (bug bz⁠#766, bug Complaints of phantom keypresses when using gfx_fb are coming in thick, and faster all the time; so far I only know of one person who actually tried an alternative backend (and said that fixed the problem).
In addition since gfx_directx is done, and is far superior to all other backends, it would be great to see whether we can use it by default, and fall back to something else if the user doesn't have DirectX 9 support. This is definitely worth delaying release for a couple days for.
\-I have not filed it on bugzilla, but NP and I ran into what appears to be another slice tree memory corruption bug while testing the A\* pathfinding scripts for his game. I just keep getting distracted.
\-Add an error reporting level selection in Custom. I'm actually very pleased with how much easier it's become to debug scripts - as long as you use the -errlvl commandline option.
\-I was going to add "Fix the current directory problem properly", although it's not very important. The problem was that the current directory is used for many different things: placing the log files, the default import directory when browsing, the export directory for screenshots, tilesets, textboxes, etc, and who knows what else. Changing the current directory was the wrong solution, instead the current directory shouldn't be used.


I've got several large things in progress in my working copy, but which I can finish up quickly.

**Comment by ohrrpgce-bugbot** [comment]
Comment author: @bob-the-hamster


> 
> > Any objections to a Friday release?
> 
> Excuuuse me, I think you mean "Any objections to a Friday release candidate?"

Sure, \*cough\* that's what I meant ;)

Actually, I literally meant a release. I feel that the wip has been MORE than stable enough to be a release for more than a month, and right now I feel like we are in danger of moving away from stability rather than towards it. (I understand that might be an irrational fear)

> Anyway, yes I do; I've been meaning to post the following, Things To Be Done
> Before Release:
> 
> -change resetgame
> -change get/setboxborder
>  (since these are breaking changes, they really should be done before the next
> stable release)

Change resetgame how?

Also, if they are breaking changes how are we going to do them in a way that doesn't break anything?

> -determine whether the default graphics backend should be changed to gfx_sdl.

Hmmm... probably.
I feel pretty good about the gfx_sdl backend right now.
In fact, I am going to make it the default for Linux right now. I am agreeable to make it the default on Windows too (depending on the gfx_directx descision)

> (bug bz⁠#766, bug Complaints of phantom keypresses when using gfx_fb are coming in
> thick, and faster all the time;

Really? I hadn't heard any. Phantom keypresses? Like randomly?

> so far I only know of one person who actually
> tried an alternative backend (and said that fixed the problem).
> In addition since gfx_directx is done, and is far superior to all other
> backends, it would be great to see whether we can use it by default, and fall
> back to something else if the user doesn't have DirectX 9 support. This is
> definitely worth delaying release for a couple days for.

That sounds cool... but doesn't that mean linking in two separate backends? Wouldn't that be a major architectural change? But you know the backends better than I do, if you think it is reasonable, I am game.

> -I have not filed it on bugzilla, but NP and I ran into what appears to be
> another slice tree memory corruption bug while testing the A\* pathfinding
> scripts for his game. I just keep getting distracted.

If you file it with the test case, I'll gladly help you track it down.

> -Add an error reporting level selection in Custom. I'm actually very pleased
> with how much easier it's become to debug scripts - as long as you use the
> -errlvl commandline option.

Yes, the new script debugging tool is frikkin' fantastic. I love it.

> -I was going to add "Fix the current directory problem properly", although it's
> not very important. The problem was that the current directory is used for many
> different things: placing the log files, the default import directory when
> browsing, the export directory for screenshots, tilesets, textboxes, etc, and
> who knows what else. Changing the current directory was the wrong solution,
> instead the current directory shouldn't be used.

I think I have fixed the worst of the side effects of changing the current directory, but not changing it would be nice too. Not a big deal releasewise for me.

> I've got several large things in progress in my working copy, but which I can
> finish up quickly.

Okay. I'll wait.

**Comment by ohrrpgce-bugbot** [comment]
Comment author: @rversteegen


> 
> > 
> > > Any objections to a Friday release?
> > 
> > Excuuuse me, I think you mean "Any objections to a Friday release candidate?"
> 
> Sure, \*cough\* that's what I meant ;)
> 
> Actually, I literally meant a release. I feel that the wip has been MORE than
> stable enough to be a release for more than a month, and right now I feel like
> we are in danger of moving away from stability rather than towards it. (I
> understand that might be an irrational fear)

Well alright, however if we do switch to gfx_sdl, then we definitely need a release candidate. Actually it's more like the other way around: a gfx_sdl release candidate would tell us whether to change the backend.

> > Anyway, yes I do; I've been meaning to post the following, Things To Be Done
> > Before Release:
> > 
> > -change resetgame
> > -change get/setboxborder
> >  (since these are breaking changes, they really should be done before the next
> > stable release)
> 
> Change resetgame how?

As promised to resolve bug bz⁠#777 (rather, feature request).

> Also, if they are breaking changes how are we going to do them in a way that
> doesn't break anything?

By keeping the existing command ids and mapping the command names to new commands in plotscr.hsd.

Well, for get/setrectborder anyway. I'm pretty sure we can get away with changing the way resetgame works, since there's currently no point using it instead of gameover in the cases for which I want to change its behaviour.

> > -determine whether the default graphics backend should be changed to gfx_sdl.
> 
> Hmmm... probably.
> I feel pretty good about the gfx_sdl backend right now.
> In fact, I am going to make it the default for Linux right now. I am agreeable
> to make it the default on Windows too (depending on the gfx_directx descision)
>
> > (bug bz⁠#766, bug Complaints of phantom keypresses when using gfx_fb are coming in
> > thick, and faster all the time;
> 
> Really? I hadn't heard any. Phantom keypresses? Like randomly?

bug bz⁠#785, bug bz⁠#753,
bug bz⁠#334, a more severe version of which was reported here: http://www.castleparadox.com/ohr/viewtopic.php?t=6450
http://rmsephy.livejournal.com/231215.html

In addition, switching backends gets around bug bz⁠#185, which is very annoying, experienced by many people, and happens 100% of the time for me on both my computers (but with different symptoms); plus bug bz⁠#766, and I thought there was another...

However, I need to get around to that testing on bug bz⁠#796, plus I've just experienced a serious bug with the pause key in gfx_sdl, which I can fix.

> That sounds cool... but doesn't that mean linking in two separate backends?
> Wouldn't that be a major architectural change? But you know the backends better
> than I do, if you think it is reasonable, I am game.

Would require a set of function pointers and a very small wrapper to dynamically load gfx_directx.dll.

**Comment by ohrrpgce-bugbot** [comment]
Comment author: @bob-the-hamster


> Well alright, however if we do switch to gfx_sdl, then we definitely need a
> release candidate. Actually it's more like the other way around: a gfx_sdl
> release candidate would tell us whether to change the backend.

Yeah, that is cool. I don't need to be as impatient as I am being :)
Can keep myself happy with safe cleanup and low-impact feature implementation to kill time until the release candidate is greenlightable. Then our testers can give gfx_sdl a serious workout.

> > Change resetgame how?
> 
> As promised to resolve bug bz⁠#777 (rather, feature request).

You misread bug bz⁠#777. It is not a feature request, and can be reproduced with absolutely no scripting whatsoever.

> > Also, if they are breaking changes how are we going to do them in a way that
> > doesn't break anything?
> 
> By keeping the existing command ids and mapping the command names to new
> commands in plotscr.hsd.
> 
> Well, for get/setrectborder anyway. I'm pretty sure we can get away with
> changing the way resetgame works, since there's currently no point using it
> instead of gameover in the cases for which I want to change its behaviour.

Might be better off with a new command? "start new game" or something like that?

> > That sounds cool... but doesn't that mean linking in two separate backends?
> > Wouldn't that be a major architectural change? But you know the backends better
> > than I do, if you think it is reasonable, I am game.
> 
> Would require a set of function pointers and a very small wrapper to
> dynamically load gfx_directx.dll.

Cool.

**Comment by ohrrpgce-bugbot** [comment]
Comment author: @rversteegen


> 
> > Well alright, however if we do switch to gfx_sdl, then we definitely need a
> > release candidate. Actually it's more like the other way around: a gfx_sdl
> > release candidate would tell us whether to change the backend.
> 
> Yeah, that is cool. I don't need to be as impatient as I am being :)
> Can keep myself happy with safe cleanup and low-impact feature implementation
> to kill time until the release candidate is greenlightable. Then our testers
> can give gfx_sdl a serious workout.
> 
> > > Change resetgame how?
> > 
> > As promised to resolve bug bz⁠#777 (rather, feature request).
> 
> You misread bug bz⁠#777. It is not a feature request, and can be reproduced with
> absolutely no scripting whatsoever.

Wow, I totally missed that, twice.

> > > Also, if they are breaking changes how are we going to do them in a way that
> > > doesn't break anything?
> > 
> > By keeping the existing command ids and mapping the command names to new
> > commands in plotscr.hsd.
> > 
> > Well, for get/setrectborder anyway. I'm pretty sure we can get away with
> > changing the way resetgame works, since there's currently no point using it
> > instead of gameover in the cases for which I want to change its behaviour.
> 
> Might be better off with a new command? "start new game" or something like
> that?

Well, I don't think I see the point. If we did, resetgame would be obsolete; so I think that's practically the same as adding a new command and remapping resetgame to it.
resetgame is pretty recent, and it's unlikely that more than 1 or 2 people has used it.

> > > That sounds cool... but doesn't that mean linking in two separate backends?
> > > Wouldn't that be a major architectural change? But you know the backends better
> > > than I do, if you think it is reasonable, I am game.
> > 
> > Would require a set of function pointers and a very small wrapper to
> > dynamically load gfx_directx.dll.
> 
> Cool.

**Comment by ohrrpgce-bugbot** [comment]
Comment author: @rversteegen

Sorry for this extended delay. Originally I thought it would be just a few days, I'm horrified that I didn't even make Christmas. I didn't expect to rewrite most of gfx_sdl; to spend a few days at family gatherings; that my menu items patch was the most horribly buggy code I've written in years, and took a number of days to fix; and so on.

However, I'm done now. And we're not waiting on gfx_directx, we're waiting on me. So I think it's time for release!

I think we should release with gfx_sdl as the default backend on all platforms, and delay a switch to gfx_directx - we could have a Ypsiliform+ rerelease when we are happy with it. I'm very confident in gfx_sdl, and don't think it needs extended testing. (I assume bug bz⁠#796 is fixed)

We could include gfx_directx.dll with Ypsiliform for people who want to try it, instructions are already in whatsnew.txt (it would be great testing!). Since there's no requirement that it work or fail gracefully on every computer, I suggest using the much smaller dll from before r3227 (160887583e2).


I decided that my original plan to change/remove get/setrectborder (from well over half a year ago) is no longer a good idea because setrectborder is actually used a fair amount. I've used it a couple times, it's in the test games, and so on. So there's no need to act before release.

The slice tree problems are such a vague problem, I'll still investigate further but it's nothing to delay release over. 

Likewise, I'd be shocked if I ever manage to reproduce Jeremy's tile corruption or bug bz⁠#738.

**Comment by ohrrpgce-bugbot** [comment]
Comment author: @rversteegen

So... what's the release plan? We've already agreed that none of the bugs marked as blockers are worth waiting around for.

**Comment by ohrrpgce-bugbot** [comment]
Comment author: @bob-the-hamster


> So... what's the release plan? We've already agreed that none of the bugs
> marked as blockers are worth waiting around for.

We'll declare tonight's nightly builds to be the Ypsiliform release candidates, and announce them on CastleParadox and SlimeSalad. If testing confirms them stable, we will branch and release in about a week. Sound okay?

**Comment by ohrrpgce-bugbot** [comment]
Comment author: @rversteegen

OK, though last time we had release candidates for over a week wasn't it because we had a new candidate every day or two? If nothing comes up, we might not need a week.

**Comment by ohrrpgce-bugbot** [comment]
Comment author: @bob-the-hamster

The release candidate threads are
http://castleparadox.com/ohr/viewtopic.php?t=6497
http://www.slimesalad.com/forum/viewtopic.php?t=3328

**Comment by ohrrpgce-bugbot** [comment]
Comment author: @bob-the-hamster

released today!

**Comment by ohrrpgce-bugbot** [comment]
Comment author: @bob-the-hamster

I will release Ypsiliform+1 when bug bz⁠#813 is fixed.

**Comment by ohrrpgce-bugbot** [comment]
Comment author: @rversteegen

I think we ought to fix bug bz⁠#812 as well. Also, I'll hunt that slice corruption bug today which I've been meaning to investigate, and see whether it exists.

I was going to suggest that we add improved script debugging to Ypsiliform+, such as line number reporting and popping open HamsterWhisper. Possibly just a couple day's work. But we could always put together Ypsiliform+2 later.

Also, about the idea to re-release Ypsiliform at some point with gfx_directx as the default on Windows: new idea, just change the default in the nightlies. That'll get it the testing it needs.

**Comment by ohrrpgce-bugbot** [comment]
Comment author: @bob-the-hamster


> I think we ought to fix bug bz⁠#812 as well. Also, I'll hunt that slice corruption
> bug today which I've been meaning to investigate, and see whether it exists.
>
> I was going to suggest that we add improved script debugging to Ypsiliform+,
> such as line number reporting and popping open HamsterWhisper. Possibly just a
> couple day's work. But we could always put together Ypsiliform+2 later.

Thats cool, but since bug bz⁠#813 is pretty crippling, I don't want to wait any longer than friday for ypsiliform+1 We can do +2 or +3 or +4 or whatever, whenever.

> Also, about the idea to re-release Ypsiliform at some point with gfx_directx as
> the default on Windows: new idea, just change the default in the nightlies.
> That'll get it the testing it needs.

That sounds groovy to me.
